File: lesson_12/structs_ts/src/stack.ts
import { ListNode } from './list_node.js';

export class Stack {
  private top: ListNode | undefined;

  constructor() {
    this.top = undefined;
  }

  push(value: number): void {
    const newNode = new ListNode(value);
    if (this.top === undefined) {
      this.top = newNode;
    } else {
      newNode.next = this.top;
      this.top = newNode;
    }
  }

  pop(): number | undefined {
    if (this.top === undefined) {
      throw new Error('Empty Stack');
    }
    const value : number = this.top.val;
    this.top = this.top.next;
    return value; 
  }

  peek(): number | null {
    if (this.top === undefined) {
      throw new Error('Empty Stack');
    }
    return this.top.val; 
  }

  isEmpty(): boolean {
    return this.top === undefined;
  }
}

File: lesson_12/structs_java/structs_app/src/main/java/com/codedifferently/lesson12/Stack.java
package com.codedifferently.lesson12;

/** Linked-list-backed stack. */
public class Stack {
  private ListNode top;

  public Stack() {
    this.top = null;
  }

  /** Push value onto the stack. */
  public void push(int value) {
    ListNode node = new ListNode(value); // assumes constructor taking (int)
    node.next = top;
    top = node;
  }

  /** Pop and return top value. Throws if empty. */
  public int pop() {
    if (top == null) {
      throw new IllegalStateException("Stack is empty");
    }
    int val = top.val; // change to top.value if your node field is named 'value'
    top = top.next;
    return val;
  }

  /** Return top value without removing. Throws if empty. */
  public int peek() {
    if (top == null) {
      throw new IllegalStateException("Stack is empty");
    }
    return top.val; // or top.value
  }

  /** Is the stack empty? */
  public boolean isEmpty() {
    return top == null;
  }
}
